### PR TITLE
Add Elasticsearch module

### PR DESCRIPTION
**Pull request by guangie88**

### CODE REVIEW

File: modules/elasticsearch/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+The actual VPC service and port are registered in `consul`. Any other services
+that require Elasticsearch service should always use the actual VPC service
+name, since the service is hosted under SSL and the SSL certificate to accept is
+registered under the VPC name (and not the `consul` service name).
+
+## Example Terraform configuration
+
+```hcl
+module "es" {
+  source = "../../../vendor/terraform-modules/modules/elasticsearch"
```
Comment (lawliet89) [inline modules/elasticsearch/README.md]: Might want to tweak this example path.

File: modules/elasticsearch/README.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+The registered `consul` service name is `elasticsearch`, and the default port
+used is `443`.
+
+The actual VPC service and port are registered in `consul`. Any other services
+that require Elasticsearch service should always use the actual VPC service
+name, since the service is hosted under SSL and the SSL certificate to accept is
+registered under the VPC name (and not the `consul` service name).
+
+## Example Terraform configuration
```
Comment (lawliet89) [inline modules/elasticsearch/README.md]: Not sure how useful the example is when most of it is `var.xxx`.  We should tweak this to be some "quick start" example instead.

- Show how we can use the outputs from `core` module using `data.terraform_remote_state`by having the data resource in the example (and `traefik` if they want to use the redirect job)
- Only show the "required" variables and some example values

File: modules/elasticsearch/cloudwatch.tf [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+  }
+}
+
+resource "aws_cloudwatch_log_resource_policy" "es_slow_index_log" {
+  policy_document = "${data.aws_iam_policy_document.es_slow_index_log.json}"
+  policy_name     = "${var.slow_index_log_name}"
+}
+
+# TODO:
```
Comment (lawliet89) [inline modules/elasticsearch/cloudwatch.tf]: Can you create an issue with the TODOs please.

File: modules/elasticsearch/main.tf [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    ebs_enabled = "${var.es_ebs_volume_size > 0 ? true : false}"
+    volume_size = "${var.es_ebs_volume_size}"
+    volume_type = "${var.es_ebs_volume_type}"
+  }
+
+  snapshot_options {
+    automated_snapshot_start_hour = "${var.es_snapshot_start_hour}"
+  }
+
+  # Note: we still have to configure Elasticsearch via the HTTP API
```
Comment (lawliet89) [inline modules/elasticsearch/main.tf]: Let's add this info to the TODO issue I mentioned above.

File: modules/elasticsearch/redirect.tf [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    az     = "${jsonencode(var.redirect_job_vpc_azs)}"
+
+    elasticsearch_service  = "${var.es_consul_service}"
+    redirect_domain        = "${local.redirect_domain}"
+    redirect_job_name      = "${var.redirect_job_name}"
+    redirect_nginx_version = "${var.redirect_nginx_version}"
+  }
+}
+
+resource "nomad_job" "redirect" {
```
Comment (lawliet89) [inline modules/elasticsearch/redirect.tf]: Let's make this job be optional.

File: modules/elasticsearch/templates/redirect.nomad [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        port = "http"
+
+        check {
+          type     = "http"
+          path     = "/"
+          interval = "10s"
+          timeout  = "2s"
+        }
+
+        tags = [
```
Comment (lawliet89) [inline modules/elasticsearch/templates/redirect.nomad]: Can mention in the README that this job will be automatically reverse proxied by Traefik if they use the Traefik module.

File: modules/elasticsearch/templates/redirect.nomad [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+          "traefik.enable=true",
+          "traefik.frontend.rule=Host:${redirect_domain}",
+          "traefik.frontend.entryPoints=internal",
+          "traefik.frontend.headers.SSLRedirect=true",
+          "traefik.frontend.headers.SSLProxyHeaders=X-Forwarded-Proto:https",
+          "traefik.frontend.headers.STSSeconds=315360000",
+          "traefik.frontend.headers.frameDeny=true"
```
Comment (lawliet89) [inline modules/elasticsearch/templates/redirect.nomad]: Let's add more tags:

```
          "traefik.frontend.headers.browserXSSFilter=true",
          "traefik.frontend.headers.contentTypeNosniff=true",
          "traefik.frontend.headers.referrerPolicy=strict-origin",
          "traefik.frontend.headers.contentSecurityPolicy=default-src 'none';",
```

**Comment by lawliet89** [comment]
Additional TODOs:
- IAM restrict who can perform ES admin
- IAM restrict who can push to ES
- Add restriction for Kibana to IP
- Use Traefik reverse proxy for ES. 

Further down the line
- Open source a Fluentd Nomad job

**Comment by guangie88** [comment]
With the new Redirect job tag changes:
`> terraform plan`

```hcl
job "kibana-redirect" {
  datacenters = ["ap-southeast-1a","ap-southeast-1b","ap-southeast-1c"]
  region      = "ap-southeast-1"
  type        = "service"

  update {
    max_parallel     = 1
    auto_revert      = true
  }

  group "kibana-redirect" {
    count = 2

    task "kibana-redirect" {
      driver = "docker"

      config {
        image = "nginx:1.14-alpine"

        port_map {
          http = 80
        }

        volumes = [
          "alloc/nginx.conf:/etc/nginx/conf.d/default.conf:ro",
        ]

        dns_servers = ["169.254.1.1"]

        # Keep default logging -- nothing special to keep track of!
      }

      template {
        data = <<EOH
server {
    listen       80;
    server_name  localhost;

    rewrite ^ https://{{ with service "elasticsearch" }}{{ (index . 0).Address }}{{ end }}/_plugin/kibana/;
}
EOH

        destination = "alloc/nginx.conf"
        change_mode = "restart"
      }

      service {
        name = "${JOB}"
        port = "http"

        check {
          type     = "http"
          path     = "/"
          interval = "10s"
          timeout  = "2s"
        }

        tags = [
          "traefik.enable=true",
          "traefik.frontend.rule=Host:kibana.locus.rocks",
          "traefik.frontend.entryPoints=internal",
          "traefik.frontend.headers.SSLRedirect=true",
          "traefik.frontend.headers.SSLProxyHeaders=X-Forwarded-Proto:https",
          "traefik.frontend.headers.STSSeconds=315360000",
          "traefik.frontend.headers.frameDeny=true"
        ]
      }

      resources {
        cpu    = 128
        memory = 64

        network {
          mbits = 200

          port "http" { }
        }
      }
    }
  }
}
```

becomes

```hcl
job "kibana-redirect" {
  datacenters = ["ap-southeast-1a","ap-southeast-1b","ap-southeast-1c"]
  region      = "ap-southeast-1"
  type        = "service"

  update {
    max_parallel     = 1
    auto_revert      = true
  }

  group "kibana-redirect" {
    count = 2

    task "kibana-redirect" {
      driver = "docker"

      config {
        image = "nginx:1.14-alpine"

        port_map {
          http = 80
        }

        volumes = [
          "alloc/nginx.conf:/etc/nginx/conf.d/default.conf:ro",
        ]

        dns_servers = ["169.254.1.1"]

        # Keep default logging -- nothing special to keep track of!
      }

      template {
        data = <<EOH
server {
    listen       80;
    server_name  localhost;

    rewrite ^ https://{{ with service "elasticsearch" }}{{ (index . 0).Address }}{{ end }}/_plugin/kibana/;
}
EOH

        destination = "alloc/nginx.conf"
        change_mode = "restart"
      }

      service {
        name = "${JOB}"
        port = "http"

        check {
          type     = "http"
          path     = "/"
          interval = "10s"
          timeout  = "2s"
        }

        tags = [
          "traefik.enable=true",
          "traefik.frontend.rule=Host:kibana.locus.rocks",
          "traefik.frontend.entryPoints=internal",
          "traefik.frontend.headers.SSLRedirect=true",
          "traefik.frontend.headers.SSLProxyHeaders=X-Forwarded-Proto:https",
          "traefik.frontend.headers.STSSeconds=315360000",
          "traefik.frontend.headers.frameDeny=true",
          "traefik.frontend.headers.browserXSSFilter=true",
          "traefik.frontend.headers.contentTypeNosniff=true",
          "traefik.frontend.headers.referrerPolicy=strict-origin",
          "traefik.frontend.headers.contentSecurityPolicy=default-src 'none';"
        ]
      }

      resources {
        cpu    = 128
        memory = 64

        network {
          mbits = 200

          port "http" { }
        }
      }
    }
  }
}
```

**Pull request merged**